package ent;
option java_package = "com.matferib.Tabuleiro.ent";

message Preco {
  optional int32 po = 1;
}

enum TipoArmadura {
  TA_LEVE = 0;
  TA_MEDIO = 1;
  TA_PESADO = 2;
}

message ArmaduraOuEscudoProto {
  optional string nome = 1;
  optional Preco preco = 2;
  optional int32 bonus = 3;
  optional int32 max_bonus_destreza = 4 [default = 100];
  optional TipoArmadura tipo = 5;
  optional int32 falha_arcana = 6;  // de 0 a 100.
  optional bool reduz_movimento = 7;
  optional int32 peso_kg = 8;
  optional int32 penalidade_armadura = 9;
  // Identificador unico, sem caracteres especiais para mapeamento em tabelas.
  optional string id = 10;
}

enum CategoriaArma {
  CAT_CAC_LEVE = 0;
}

enum TipoDano {
  TD_CORTANTE = 0;
  TD_CONCUSSAO = 1;
  TD_PERFURANTE = 2;
}

// O medio vem de tabela e os outros gerados automaticamente.
message StringPorTamanho {
  optional string medio = 1;
  optional string pequeno = 2;
  optional string grande = 3;
}

message ArmaProto {
  optional string id = 1;
  optional string nome = 2;
  repeated CategoriaArma categoria = 3;
  optional string preco = 4;
  optional StringPorTamanho dano = 5;
  // Presente em armas duplas.
  optional StringPorTamanho dano_secundario = 6;
  optional string critico = 7;
  optional StringPorTamanho peso = 8;
  repeated TipoDano tipo_dano = 9;
  optional int32 incremento_quadrados = 10;
}

message TabelaArmaduraOuEscudo {
  repeated ArmaduraOuEscudoProto armaduras = 1;
  repeated ArmaduraOuEscudoProto escudos = 2;
}

message TabelaArmas {
  repeated ArmaProto armas = 1;
}

// O arquivo tabelas.asciiproto eh desse tipo.
message TodasTabelas {
  optional TabelaArmaduraOuEscudo tabela_armaduras = 1;
  optional TabelaArmaduraOuEscudo tabela_escudos = 2;
  optional TabelaArmas tabela_armas = 3;
}
